keepalived/core/keepalived_netlink.c

```diff
@@ -1127,7 +1127,10 @@ netlink_if_address_filter(__attribute__((unused)) struct sockaddr_nl *snl, struc
 							 vrrp->ifp) &&
 						 vrrp->family == ifa->ifa_family &&
 						 vrrp->saddr.ss_family != AF_UNSPEC &&
-						 (!vrrp->saddr_from_config || is_tracking_saddr)) {
+						 (!vrrp->saddr_from_config || is_tracking_saddr) &&
+						 (vrrp->family != AF_INET6 || (vrrp->family == AF_INET6 &&
+						  (vrrp->dont_track_primary && IF_ISUP(ifp))) ||
+						  !vrrp->dont_track_primary)) {
 						down_instance(vrrp);
```

Comment:
This change can be re-written as:
```
                                                         vrrp->ifp) &&
                                                 vrrp->family == ifa->ifa_family &&
                                                 vrrp->saddr.ss_family != AF_UNSPEC &&
                                                 (!vrrp->saddr_from_config || is_tracking_saddr) &&
                                                 (vrrp->family == AF_INET || 
                                                  !vrrp->dont_track_primary ||
                                                  IF_ISUP(ifp))) {
```

So the change to the code is that if the VRRP instance is IPv6 and dont_track_primary is set and the interface is down, then the VRRP instance is not downed.

The trouble with this is that this is triggered by the source address that the VRRP instance is using has been deleted, and we cannot allow keepalived to be sending from an address that no longer exists (and might even have been moved to another system) - my recollection is that for IPv6 the kernel does not allow sending from an IP address that is not configured (but I may be wrong here).

Reply:
With my patch, I get that on IPv6. No crash. Same behaviour on IPv4 dont_track_primary.

Apr 26 11:02:18 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:19 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:20 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:21 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:22 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:23 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:24 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:25 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)
Apr 26 11:02:26 ubuntu Keepalived_vrrp[1802]: (vrrp2): send advert error 22 (Invalid argument)

Reply:
@louis-oui I would expect the above error with your patch, since keepalived is attempting to send an advert on eth5 with a source address that is not configured. From testing it appears that EINVAL is returned both when the source address is not configured and also when the interface is administratively down.

Reply:
@pqarmitage I confirm 'send advert error 22 (Invalid argument)' is coming from this Ipv6 patch